docs/oppen_algorithm.md

```diff
@@ -0,0 +1,359 @@
+# Derek C. Oppen's pretty pritning algorithm
+
+Derek C. Oppen's algorithm provides an efficient solution for formatting and printing code with clean indentation that respects a specified screen width.
+
+When working with large blocks of code, ensuring proper indentation and alignment is crucial for readability and maintainability. However, achieving this while adhering to a specific line width can be a challenging task.
+
+This algorithm is great if you are working with code, documentation, logging, or any situation that requires pretty-printing.
+
+## The printing tokens
+
+Oppen's algorithm expects a list of tokens describing the structure of the expected output.
+
+The main tokens are:
+
+|   Name    | Description
+|-----------|------------
+| STRING    | String literal to be displayed.
+| BEGIN     | Opens a group.
+| END       | Closes a group.
+| BREAK     | Potential line break (adds a new line if the text doesn't fit on the line).
+| LINEBREAK | Guaranteed line break.
+
+A group can contain any number of groups, `String`, `Break`, and a `LineBreak`. Groups exist to containerize different _groups_ of tokens with respect to line breaking.
+
+Containerizing groups allows the creation of subsets that are independent from other subsets, and semi-independent from their parent subsets.
+
+Example for a maximum desired printing width of 3:
+```
+[
+  BEGIN#1, # Start of group 1
+
+  STRING = 'a',
+
+  BREAK#1,
+  STRING = 'b',
+
+  BREAK#2,
+  STRING = 'c',
+
+  END#1,   # end of group 1
+
+  EOF
+]
+
+=>
+a
+b
+c
+```
```

Comment:
````suggestion
```
[BEGIN, STRING(a), BREAK, STRING(b), BREAK, STRING(c), END]
```

We just defined a group to print a line of width 5: 3 characters for the `STRING` tokens, and 2 for the `BREAK` which are literal spaces by default.

Since our desired line width is 3, we need to break the group. `BREAK`s are transformed into new lines, and we get:
```
a
b
c
```
````